File: 0017-Letter-Combinations-of-a-Phone-Number/lc0017.js
// JavaScript solution.

/**
 * @param {string} digits
 * @return {string[]}
 */

var letterCombinations = function(digits) {
    let sz = digits.length;
    let ans = [];
    let res = Array(sz).fill('0');
    let dig =  ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz'];
    if (sz === 0) { return ans; }

    function backtracking(depth) {
        if (depth === sz) {
            ans.push(res.join(""));
            return;
        }

        let button = digits[depth];
        for (let letter of dig[button]) {
            res[depth] = letter;
            backtracking(depth+1);
        }
    }

    backtracking(0);
    return ans;
};

File: 0017-Letter-Combinations-of-a-Phone-Number/lc0017.rs
// Rust solution.

struct Bcparam<'a> {
    digits: &'a str,
    ans: Vec<String>,
    res: Vec<char>,
}

impl<'a> Bcparam<'a> {
    #[inline]
    pub fn dig() -> Vec<&'static str> {
        vec!["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
    }
}

impl Solution {
    pub fn letter_combinations(digits: String) -> Vec<String> {
        let sz = digits.len();
        if (sz <= 0) { return vec![]; }
        let mut params = Bcparam {
            digits: &digits[..],
            ans: Vec::new(),
            res: vec!['0'; sz],
        };
        Solution::backtracking(0, sz, &mut params);
        params.ans
    }

    pub fn backtracking(depth: usize, sz: usize, params: &mut Bcparam) {
        if depth == sz {
            params.ans.push(params.res.iter().collect::<String>());
            return;
        }

        let button = (params.digits.as_bytes()[depth] as char).to_digit(10).unwrap();
        for letter in Bcparam::dig()[button as usize].chars() {
            params.res[depth] = letter;
            Solution::backtracking(depth+1, sz, params);
        }
    }
}
